prisma-fmt/src/offsets.rs

```diff
@@ -0,0 +1,134 @@
+use lsp_types::{Position, Range};
+use psl::{diagnostics::FileId, parser_database::ast::Span};
+
+/// The LSP position is expressed as a (line, col) tuple, but our pest-based parser works with byte
+/// offsets. This function converts from an LSP position to a pest byte offset. Returns `None` if
+/// the position has a line past the end of the document, or a character position past the end of
+/// the line.
+pub(crate) fn position_to_offset(position: &Position, document: &str) -> Option<usize> {
+    let mut offset = 0;
+    let mut line_offset = position.line;
+    let mut character_offset = position.character as i64;
+    let mut chars = document.chars();
+
+    while line_offset > 0 {
+        loop {
+            match chars.next() {
+                Some('\n') => {
+                    offset += '\n'.len_utf8();
+                    break;
+                }
+                Some(chr) => {
+                    offset += chr.len_utf8();
+                }
+                None => return Some(offset),
+            }
+        }
+
+        line_offset -= 1;
+    }
+
+    while character_offset > 0 {
+        match chars.next() {
+            Some('\n') | None => return Some(offset),
+            Some(chr) => {
+                offset += chr.len_utf8();
+                character_offset -= chr.len_utf16() as i64;
+            }
+        }
+    }
+
+    Some(offset)
+}
+
+#[track_caller]
+/// Converts an LSP range to a span.
+pub(crate) fn range_to_span(range: Range, document: &str, file_id: FileId) -> Span {
+    let start = position_to_offset(&range.start, document).unwrap();
+    let end = position_to_offset(&range.end, document).unwrap();
+
+    Span::new(start, end, file_id)
+}
+
+/// Gives the LSP position right after the given span.
+pub(crate) fn position_after_span(span: Span, document: &str) -> Position {
+    offset_to_position(span.end - 1, document)
+}
```

Comment:
This was updated in https://github.com/prisma/prisma-engines/pull/4870

Reply:
the changes from that PR probably also needs to be iterating over bytes and not chars

Reply:
This is my bad. I fixed it. Since `Span` is calculated using byte offsets, there was no need to do the additional fix.